### DISCUSSION TITLE
Please, one, only one quickstart for Django

### DISCUSSION THREAD

**Discussion by RicardoFernandez-UY** [keycloak/keycloak#13510] [Q&A]
Hi,
I hope your are fine!
All examples, quickstarts, and everything else everywhere is oriented toward Spring, SpringBoot, Angular, Nodejs, etc.
And, that's fine!
But, why not at least one good working example for Django? It will undoubtedly be appreciated!!
Thank you!! and best regards!

**Comment by abstractj** [comment]
@RicardoFernandez-UY thank you for your interest. Spring, SpringBoot, Angular.js and Node.js adapters will soon be EOL on Keycloak. See: https://www.keycloak.org/2022/03/releases.html. 

Our team would love to provide support and examples to more language, but considering our bandwidth and expertise, we are unable to provide examples with Django. If this is something that you have the knowledge, time and would like to contribute that would be welcome.
Searching for it, I could find as a starting point https://django-keycloak.readthedocs.io/en/latest/.

I hope it helps.

**Reply by RicardoFernandez-UY** [reply]
@abstractj  Thank you very much for your answer !! and I'm really sorry for the duplication of issues.
If I had the knowledge, you can be sure I would have already contributed that adapter to the project. Unfortunately, that is not the case; I'm a newbie in the programming world.

I understand you are under a lot of changes in pursuit of a better future, and I applaud that.!
On the dark side, I can't feel secure with the security tool (what a paradox), as there is too much information out there, much of which is contradictory and old, and if I stay with the official version, I'm unable to use the tool, because there is no adapter and no information on how to work around that fact.
I have recently discovered the Keycloak Gatekeeper and I felt great relief as it seemed to be the solution to my problem; but it lasted like 5 minutes, as I later discovered it was already deprecated or in its EOL process as you like to call it. And now there's a sort of spinoff called OAuth2-Proxy. (this happens like 20 minutes ago, so I still couldn't investigate it). 
Are you aware of this one? If so, would you recommend it?

So, I hope you understand why I'm so baffled and stuck, discovering new things at every moment but being unable to use any of them; it seems too hard to land on stable ground. Everything is pushing me to end using the internal Django security system ... but, I will not give up yet :-), and I'll work a bit more on trying to integrate KC into my project.

I appreciate the time you took to find django-keycloak library. Unfortunately, I've already checked it a while ago and it's been unmaintained for like 4 years and it doesn't work anymore with newer versions of the products.

In summary, thank you very much for answering my question and for your time, I do appreciate it a lot!
I feel sad I still can't become confident in using KC for securing my environment, because really I think it's a great tool.

I wish you the best!!
Warm regards!!

**Reply by abstractj** [reply]
Hey @RicardoFernandez-UY no worries and I completely understand, your feedback is welcome, and there's always a room for improvements. 

Regards your question about Gatekeeper, the project became superseded by oauth2-proxy (https://github.com/oauth2-proxy/oauth2-proxy), which has successful adoption, and many contributors around it. I'd recommend the project, if you don't manage to find a good adapter for Django.

@JoelSpeed is one of the creators/maintainers behind the project, and they also provide some instructions about the usage of Keycloak with OAuth2 Proxy (https://oauth2-proxy.github.io/oauth2-proxy/docs/configuration/oauth_provider/#keycloak-auth-provider). Might be worth to give it a try. Checking their [contrib](https://github.com/oauth2-proxy/oauth2-proxy/tree/master/contrib/local-environment) folder, there's a dedicated section for Keycloak.

I'd also check the suggestions made by @Ali-Javanmardi

**Reply by RicardoFernandez-UY** [reply]
Hi @abstractj thanks again for the follow-up! 
I'll certainly check all the materials you mentioned and also what @Ali-Javanmardi recommended.
That will surely keep me busy for a while :-).

**Comment by Ali-Javanmardi** [comment]
If all you want is ability to use oauth with your Django project and you do not want control user data in Keycloak by your Django code, I may suggest you django-allauth :
https://github.com/pennersr/django-allauth

I have tried it sometimes ago and found it straightforward and easy to integrate many oauth providers with Django.

but in case you want to manage your Keycloak data from your Django  code, I suggest you have a look on python-keycloak :
https://github.com/marcospereirampj/python-keycloak

**Reply by RicardoFernandez-UY** [reply]
Hi,
Thank you very much for your answer!
I would like if possible, to delegate all security in Keycloak, that is, define roles, groups, users, etc in Keycloak with the GUI provided by Keycloak, and also store all the user information there. (the exact opposite of what would happen if I implemented security using Django native auth capabilities)
I don't want to interact with KC from Django code. In my Django code, all I want to do is to use the permissions/roles that are defined in KC, and should be passed back in the tokens once you log in, to protect resources in the application (data, functionality, etc.)

So according to what you said, django-allauth would be the way to go? I've heard of it a while ago, but I understood that it was a set of Django apps that implemented something similar to what KC provides and also integrated social authentication (maybe I misunderstood, I will review it again)

**Reply by Ali-Javanmardi** [reply]
Django-allauth will let you login to your Django app by using an external authentication system, but after user logged in will create a user data record inside your Django database and so you can manage his/her access control in your Django code.

However as you mentioned you want to store all user profile related stuff in KC, allauth do not what you are looking for.

In my case I tried to use django-oauth-toolkit instead of allauth but after some while as I wanted to manage more user data by KC and it was not a good idea to embed everything in one jwt token( a token with all data is bloated and if you transfer it by every request make a huge network overhead) I decided to rewrite my own login middleware in Django code.

To do that I used python-keycloak for my connection to keycloak and also used PyJWT to read and parse jwt token received by request.
I'm not sure if the way I did it is a good secure solution and so I hesitate to share its code here. But you may read documentations related to those libraries and decide yourself.

**Reply by RicardoFernandez-UY** [reply]
Hi Ali,
Well, I see you have traveled quite a path in this area!
To be honest, at present I'm at the stage of having a big mess in my head regarding what to do.

I don't expect to store too much information regarding user profile (at least not about basic stuff like permissions, roles, and that sort of thing), but I understand your point, and I guess it also might depend on how much traffic you have in your app. 
I remember having read something about caching tokens on the Django side to avoid that problem you mentioned, or selectively choosing what information to send from the Keycloak side in a given token, but I can't tell if is a good practice.

What I would like to avoid it to have to code from scratch because I'm not a very good programmer yet.
On the other hand, taking a look at a code written by a more seasoned programmer like you would probably help me understand how a full coded solution should look like, and probably also to better understand how to work with the JWTs  inside the application
So, first I will review these libraries you mentioned and do some testing.
Regarding your hesitation about sharing that code, if there is a way to share it privately I would appreciate it (as long as doing it does not brake any site rule), but please do not any feel kind of pressure about it.

Thank you very much!!

**Reply by Ali-Javanmardi** [reply]
As I mentioned I'm not sure how secure is my own code, so If you are looking for a practice or code sample from professional people  it will be much better to look for open source repositories that have a lot of contributors and negotiations. For example django-allauth and django-oauth-toolkit are good samples.

**Comment by codespearhead** [comment]
As others have mentioned, the issue isn't the lack of a Keycloak quickstart for your framework, but rather the scarcity of OIDC clients in many programming languages [1], let alone integrations between those clients and your chosen framework.

Maintaining an OIDC client is very demanding mainly because the big players are slow to make their servers OIDC-compliant [2]. Hence, they tend to provide their own SDKs for multiple programming ecosystems, then neglect most of them according to business demand. Since the C# community prefer proprietaty solutions and the Java one prefer to have stuff standards-compliant, they're the ones that tend to stick: the proprietary C# SDK, and the OIDC client provided by whatever Java framework you're using.

And that's precisely why the OIDC client implementations listed in [1] tend to be feature-incomplete and are generally deprecated.

Fun fact: as of today this [3] seems to be the only business that provides an OIDC client for Wordpress, hence they're contracted left and right because there doesn't seem to be any other production-ready alternatives. That reminds me of Amit Agarwal, who makes crazy money by cateting to a niche [4][5].

By the way, I'm seriously considering developing an open-source OIDC client in C with official Python bindings, as I need to use Python at work. Creating bindings is far easier and more maintainable than reimplementing the complex core logic in multiple languages. This approach would allow other languages to simply create bindings rather than duplicating efforts. If anyone is interested in this idea, feel free to comment below.

***

[1] https://openid.net/developers/certified-openid-connect-implementations/ (see "Certified Relying Party Libraries" section)
[2] https://github.com/ramosbugs/openidconnect-rs/issues/122
[3] https://plugins.miniorange.com/wordpress-sso
[4] https://www.linkedin.com/posts/shaanpuri_amit-agarwal-is-a-one-man-team-making-20m-activity-6978796033640124416-rL-l/
[5] https://www.labnol.org/about/
